File: backend/apps/zones/charges.py
import math

from .models import Zone


def round_half_up(value: float) -> int:
    """Match JS Math.round (round half away from zero for .5)."""
    return int(math.floor(value + 0.5))


def find_zone_by_district(district: str) -> Zone | None:
    active = list(Zone.objects.filter(is_active=True))
    for z in active:
        if district in (z.districts or []):
            return z
    return active[0] if active else Zone.objects.first()


def compute_charge(zone: Zone, delivery_type: str, weight: float, cod_amount: float) -> dict:
    """Mirror lib/charges.ts computeCharge.

    delivery = base(zone, type) + 10 per kg over 1kg
    cod      = max(10, cod_amount * cod_charge_percent / 100), or 0 when no COD
    """
    base = zone.express_charge if delivery_type == "express" else zone.regular_charge
    weight_extra = max(0, math.ceil((weight or 0) - 1)) * 10
    delivery_charge = round_half_up(base + weight_extra)
    if cod_amount and cod_amount > 0:
        cod_charge = max(10, round_half_up(cod_amount * zone.cod_charge_percent / 100))
    else:
        cod_charge = 0
    return {
        "zoneName": zone.name,
        "deliveryCharge": delivery_charge,
        "codCharge": cod_charge,
        "totalCharge": delivery_charge + cod_charge,
    }

File: frontend/lib/charges.ts
import { getZones } from "@/lib/zone-store";
import type { Zone, DeliveryType } from "@/types";

// Resolve the delivery zone for a recipient district. Falls back to the first
// active zone (Dhaka City) when the district isn't explicitly mapped. Returns
// undefined while zones are still loading from the API (empty store).
export function findZoneByDistrict(district: string): Zone | undefined {
  const allZones = getZones();
  const match = allZones.find(
    (z) => z.isActive && z.districts.includes(district),
  );
  return match ?? allZones.find((z) => z.isActive) ?? allZones[0];
}

export interface ChargeBreakdown {
  zoneName: string;
  deliveryCharge: number;
  codCharge: number;
  totalCharge: number;
}

// Mirrors the M6 backend compute_charge utility so the preview shown at booking
// time matches what the API will persist later.
//   delivery = base(zone, type) + weightExtra (10৳ per kg over 1kg)
//   cod      = max(10, codAmount * zone.codChargePercent / 100)
export function computeCharge(
  zone: Zone | undefined,
  deliveryType: DeliveryType,
  weight: number,
  codAmount: number,
): ChargeBreakdown {
  // Zones still loading (empty store) — return a zero breakdown so the preview
  // renders "৳0" instead of crashing; it fills in once zones arrive.
  if (!zone) {
    return { zoneName: "", deliveryCharge: 0, codCharge: 0, totalCharge: 0 };
  }
  const base =
    deliveryType === "express" ? zone.expressCharge : zone.regularCharge;
  const weightExtra = Math.max(0, Math.ceil(weight - 1)) * 10;
  const deliveryCharge = Math.round(base + weightExtra);
  const codCharge =
    codAmount > 0
      ? Math.max(10, Math.round((codAmount * zone.codChargePercent) / 100))
      : 0;
  return {
    zoneName: zone.name,
    deliveryCharge,
    codCharge,
    totalCharge: deliveryCharge + codCharge,
  };
}
